cypress-io/cypress
# fix: same site cookie context and duplicate cookies

- Closes #23132

### User facing changelog
Fixes an issue introduced in `10.3.0` and further exposed in `10.4.0` that omitted same-site cookies when URL Scheme, Domain, and Top Level Domain match, but the ports are different (same-site).

### Additional details
In #22320 we added improved cookie handling that handles cookie management in the proxy server. When we calculate `SameSiteContext`, we were using `SameOriginPolicy` to calculate the context of the request. This is a bit different than `SameSite`, which cookies use to determine context as subdomains and ports do not need to match to [be considered same-site](https://httpwg.org/http-extensions/draft-ietf-httpbis-rfc6265bis.html#section-1-3). 

In #22963 we removed the [isAUTFrame](https://github.com/cypress-io/cypress/pull/22963/files#diff-525d773dc9e405242dbb041d386d3b3e6c6c07bb02bb3b4436fb96de5cdb9072L410) check in `checkIfNeedsCrossOriginHandling`. In the case of https://github.com/cypress-io/cypress/issues/23132#issuecomment-1206842968, I think `isAUTFrame` used to evaluate to `false`  for the fetch/XHR requests being made, which leveraged `Set-Cookie` over `X-Set-Cookie`. After `10.4.0`, the check was removed and `X-Set-Cookie` was being used as it believed the requests needed to handle the cookies in a third party context, when in fact they did not. 

Until #23551 is implemented, we should likely use `Set-Cookie`, even if the setting of the actual cookie fails, as we currently cannot discern the difference between requests that are specified in the browser with [XMLHttpRequest Credentials](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/withCredentials) or [fetch credentials](https://developer.mozilla.org/en-US/docs/Web/API/fetch#credentials) and once that aren't (or use more strict fetch credentials like `same-origin` or `omit`).


### Steps to test
Unit tests were added for functions `getSameSiteContext` in `@packages/proxy` and `urlOriginsMatch`  in `packages/network` to verify the behavior is what we expect.

Unit tests for verifying duplicate cookie behavior and integration tests with tough-cookie were also added to make sure we integrate correctly.

E2E tests were added in `cookie_behavior.cy.ts` to serve as a base for when we explore #23551 to make sure the correct spec behavior is implemented. There are several tests with `FIXME`s as well as comments to what the assertions should be. In the current context, we may optimistically apply cookies to request, as if each request was called `withCredentials` or with `include`d credentials, which the tests should demonstrate. 

### How has the user experience changed?


### PR Tasks
